docs/tekton-bundle-contracts.md

```diff
@@ -0,0 +1,97 @@
+<!--
+---
+linkTitle: "Tekton Bundle Contracts"
+weight: 8
+---
+-->
+# Tekton Bundle Contract v0.1
+
+When using a Tekton Bundle in a task or pipeline reference, the OCI artifact backing the
+bundle must adhere to the following contract.
+
+## Contract
+
+Only Tekton CRDs (eg, `Task` or `Pipeline`) may reside in a Tekton Bundle used as a Tekton
+bundle reference.
+
+Each layer of the image must map 1:1 with a single Tekton resource.
```

Comment:
This is concerning because there are some performance implications and some practical limitations with respect to layer sizes. If you map 1:1 resources to layers, you will have some practical upper bound for how many resources are reasonable to bundle together.

You'll incur at least one HTTP request per layer when fetching these.

If for some reason these get pulled by a runtime, there's a limit to the number of layers that union file systems can support.

Registries will have an upper limit to the size of a manifest, so keep that in mind.

Reply:
That's good context and I appreciate the info. On the one hand, I haven't proposed or written out a limit so there could be abuse. On the other hand, etcd imposes a limitation on the size of each resource (therefore each layer). There is nothing stopping someone from trying to create an image with 100 objects I suppose but as you mention, the limit is registry dependent so perhaps this can be handled with user messaging on the side of the tool rather than the contract?

Reply:
I don't have enough background to know if 100 objects is reasonable or not.

I'm curious what the reasoning is for having these each as one layer. Do you need the layer annotations to be able to parse this correctly?

I do like that this is simple and doesn't required gzip/tar to create or consume, but I'm wondering if it makes sense to have some kind of escape hatch for a bunch that _does_ need a lot of files. IMO, that would be a good use case for `application/tar`, which could contain an arbitrary number of resources. Again, though, I don't have the context to know if that makes any sense.

Another thing I'd worry about with a large number of layers is tail latency -- this gets a lot worse as you increase the number of layers, which you really don't want in your controller.